KAFKA-3989: initial support for adding a JMH benchmarking module.

bbejeck opened the pull request:
This is related to KAFKA-3973, as the benchmarks were re-written in JMH as part of this PR. Results are posted in the KAFKA-3973 jira ticket.

bbejeck committed:
KAFKA-3989: Added support for adding a JMH benchmarking module, includes shadow plugin, jmh.sh script for running benchmarks.

KAFKA-3989: Updated README file to be more complete, added how to capture output in file, removed comment about adding jmh-gradle plugin.

KAFKA-3989: Small changes to readme file.

KAFKA-3989: updated the options section of README file, added jmh.sh script to aid in running benchmarks

bbejeck commented:
ping @ijuma @enothereska @mjsax

dguy commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/streams/MockMemoryLRUCache.java:
```diff
@@ -0,0 +1,275 @@
+
+            //just for testing would limit cache size to ~ 2GB
+            @Override
+            public int size() {
```
Curious as to why you override `size`? Can't you just use `totalMemoryUsed` in  `removeEldestEntry`? Just seems a bit strange to morph `size` into something other than its intended meaning.

bbejeck commented:
@ijuma @enothereska 

Could you please review when you have a moment?

ijuma commented on jmh-benchmarks/README.md:
````diff
@@ -0,0 +1,196 @@
+To get all the jmh options you run jmh with the -h flag or just look below:
+
+```text
+java -jar jmh-benchmarks/build/libs/kafka-jmh-benchmarks-0.10.1.0-SNAPSHOT-all.jar -h
````
It seems like this will get out of date pretty quickly. How about we include examples for the common cases and provide the command for people who want to get the full list?

ijuma commented on jmh-benchmarks/README.md:
```diff
@@ -0,0 +1,196 @@
+     iterations (again a default value).
+
+* By default all output goes to stdout.  If it is desired to capture the output in a file add -o&lt;filename&gt; to the jmhArgs property:
+   `./gradlew jmh-benchmarks:jmh -DjmhArgs=-f2,-oMemBytesBenchmark.txt,MemoryBytesCacheBenchmark`
```
Passing parameters via -DjmhArgs seems a bit clunky compared to `https://github.com/melix/jmh-gradle-plugin` although it's nice to be able to use plain JMH arguments.

bbejeck replied:
@ijuma  thanks for the comments.

 I agree using -DjmhArgs is clunky.  But in the gradle-jmh-plugin all JMH parameters are configured in a static block in the build.gradle file.  If you want/need different configs for individual benchmarks then variables in that block need to be changed.  (It seems that command line options in the jmh-gradle-plugin are an unresolved issue https://github.com/melix/jmh-gradle-plugin/issues/46, https://github.com/melix/jmh-gradle-plugin/issues/54)

I have an alternante proposal.  What do you think about a shell script that:
1. Accepts plain JMH arguments;  "./jmh -o outPutfile.txt -f 10 MyBenchmarks"
2. Delegates to gradle to build the shadow jar file.
3. Runs JMH from command line with args passed in step 1.

ijuma replied:
That sounds good to me.

bbejeck commented:
@ijuma 
updated per comments and included an initial cut at including a bash script.

bbejeck commented:
rebased this PR

bbejeck commented:
@ijuma 

Sorry for being a pest, can you take a look ? I'm guessing this won't be merged until KAFKA-3776 is completed, is that correct?

ewencp commented on jmh-benchmarks/jmh.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
This fails checkstyle due to missing the license header. Can you add one similar to the files in the `bin/` directory?

ewencp commented on jmh-benchmarks/jmh.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+
+gradleCmd="../gradlew"
```
This could be changed to find the current directory of the script instead, which would avoid dependence on running from a specific location.

ewencp commented on jmh-benchmarks/jmh.sh:
```diff
@@ -0,0 +1,18 @@
+
+
+gradleCmd="../gradlew"
+libDir="build/libs"
```
Same here wrt directories relative to the working directory.

ewencp commented on build.gradle:
```diff
@@ -766,6 +771,43 @@ project(':streams:examples') {
   }
 }
 
+project(':jmh-benchmarks') {
```
Should this declaration exist twice in the script? Why do we need the separate invocation above that just does the `apply plugin` and a separate one here with everything else? Why not merge them?

ewencp commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/streams/MemoryBytesCacheBenchmark.java:
```diff
@@ -0,0 +1,82 @@
+
+    MockMemoryLRUCache<String, String> memoryCache;
+    MockMemoryLRUCache<String, String> bytesCache;
+    String key = "the_key_to_use";
```
These should be `private final`

ewencp commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/streams/MockMemoryLRUCache.java:
```diff
@@ -0,0 +1,275 @@
+/**
```
Is there a reason we need this instead of `org.apache.kafka.common.cache.LRUCache`?

bbejeck replied:
@ewencp - I did this so I could inject the org.github.jamm.MemoryMeter object for benchmarking  tracking cache size by memory.

ewencp reviewed (commented):
I appreciate the inclusion of some per benchmarks, seems like a good ideal overall. I have a few questions, but overall this looks great. If we could even get a reasonable subset that run in a smaller fraction of time, I'd love to get some of these tests as part of the standard suite of tests run for every merge.

ijuma commented:
Sorry for the huge delay getting to this @bbejeck. Thankfully @ewencp has jumped in.

bbejeck committed:
KAFKA-3989: updates per comments, renamed jar file to remove version number

bbejeck commented:
@ewencp thanks for the review. Updates made per comments, although I'm not sure the changes for the working directory are what you had in mind, just let me know.

EDIT:
I'm coupling the generation of the jmh uber jar/benchmark execution  in one script, with the idea that the benchmarks will always use the latest code.  This is why I'm relying on the execution directory location being the Kafka project directory or the jmh-benchmarks subproject directory.

Would it better to break out into two scripts:
1. A script that executes the benchmarks and can be kicked from any directory and would not rely on the ./gradlew call for building the uber jar.
2.  A second script that combines gradle uber jar creation and benchmark execution thus requiring the execution location being in Kafka project or jmh-benchmark subproject.

@ijuma no worries! thanks for your support on this PR!

bbejeck commented:
@ewencp
 I know things are super busy now, but when you have a moment can you look over the updates?
Thanks!

bbejeck commented:
rebased PR

@ewencp  @ijuma   thanks for the reviews so far, just sending a tickler for  0.10.3

ijuma commented on build.gradle:
```diff
@@ -27,6 +27,7 @@ buildscript {
     classpath "org.ajoberstar:grgit:1.7.0"
     classpath 'com.github.ben-manes:gradle-versions-plugin:0.13.0'
     classpath 'org.scoverage:gradle-scoverage:2.1.0'
+    classpath 'com.github.jengelman.gradle.plugins:shadow:1.2.3'
```
We can bump this to 1.2.4.

bbejeck replied:
ack

ijuma commented on build.gradle:
```diff
@@ -767,6 +768,49 @@ project(':streams:examples') {
+      compile project(':streams')
+      compile 'org.openjdk.jmh:jmh-core:1.13'
+      compile 'org.openjdk.jmh:jmh-generator-annprocess:1.13'
+      compile 'org.openjdk.jmh:jmh-core-benchmarks:1.13'
```
We can bump these to 1.17.5.

bbejeck replied:
ack

ijuma commented on jmh-benchmarks/README.md:
```diff
@@ -0,0 +1,63 @@
+* By default all JMH output goes to stdout.  To run a benchmark and capture the results in a file:
+`./jmh.sh -f 2 -o benchmarkResults.txt MemoryBytesCacheBenchmark`
+
+
```
Extra line?

bbejeck replied:
ack

ijuma commented on jmh-benchmarks/README.md:
````diff
@@ -0,0 +1,63 @@
+```
+To view all options run jmh with the -h flag. 
+
+
````
Extra line.

bbejeck replied:
ack

ijuma commented on jmh-benchmarks/jmh.sh:
```diff
@@ -0,0 +1,44 @@
+libDir="${base_dir}/build/libs"
+
+
+
```
Extra lines.

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/streams/MemoryBytesCacheBenchmark.java:
```diff
@@ -0,0 +1,83 @@
+
+        bytesCache = new MockMemoryLRUCache<>("trackBySizeBytes", 100000, Serdes.String(), Serdes.String());
+        bytesCache.setIsCachingBytes(true);
+    }
```
Generally, I think it's a better practice to use state classes to clearly separate state to be used by the benchmark  (and ensure we don't end up with some shared state that can affect benchmark results). Could we do that here?

ijuma commented on jmh-benchmarks/src/main/java/org/apache/kafka/jmh/streams/MockMemoryLRUCache.java:
```diff
@@ -0,0 +1,275 @@
+ *
+ * @see org.apache.kafka.streams.state.Stores#create(String)
+ */
+public class MockMemoryLRUCache<K, V> implements KeyValueStore<K, V> {
```
Do we really need this class? The way I see it, we should only include benchmarks for production code in the AK repo. For other one-offs, we should do it in separate repositories so that we don't have to maintain it going forward. If we just need an example benchmark in the source code until we have real ones, maybe we could do something simpler that doesn't require a 275 lines class?

bbejeck replied:
Probably not.  I had initially constructed something to test tracking objects by size in memory vs. a straight count of entries.  But you are right about this being overkill, so I have replaced it with a much simpler example using existing AK production code.

ijuma reviewed (commented):
Sorry for the delay. I took a pass and left some comments. I'll make sure to allocate time to merge this ASAP if the comments are addressed.

bbejeck commented:
@ijuma thanks for the review comments, I'll address very soon.

ijuma commented on jmh-benchmarks/README.md:
```diff
@@ -0,0 +1,63 @@
+###JMH-Benchmark module
```
One more thing: do we want to call this `jmh-benchmarks` or simply `benchmarks`? I can see pros/cons to either option, but interested in your thoughts.

bbejeck replied:
I could go either way as well, but I'm leaning towards using `jmh-benchmarks` vs. `benchmarks` to re-enforce the fact we only should use JMH for benchmarking.  Having said that if you prefer `benchmarks`, I'm fine with that.

ijuma replied:
OK to leave it like it is for now. We can rename it later if we decide it's better.

bbejeck committed:
KAFKA-3989: changes per comments
                       updated version of JMH
                       updated version of gradle shade plugin
                       removed previous benchmark with simpler example

bbejeck committed:
KAFKA-3989: update to the correct JMH version, README.md tweaks

bbejeck commented:
@ijuma thanks for the review.

I've addressed the comments and most notably have removed the old benchmark class and replaced it with a much simple one using the o.a.k.common.cache.LRUCache for an example benchmark.

EDIT: also rebased PR.

bbejeck commented:
my fault - addressing checkstyle failures now

bbejeck committed:
KAFKA-3989: fixed license header

ijuma commented on jmh-benchmarks/jmh.sh:
```diff
@@ -0,0 +1,44 @@
+
+if [ ${base_dir} == "." ]; then
+     gradlew_dir=".."
+elif [ ${base_dir} == ${jmh_project_name} ]; then
```
I noticed that this doesn't work if one does:

```sh
./jmh-benchmarks/jmh.sh
```

bbejeck replied:
@ijuma thanks, I'll take a look and issue a follow up PR

ijuma commented:
Thanks for your patience, merged to trunk. We can improve things further in subsequent PRs.

bbejeck commented:
@ijuma, @ewencp  Thanks to both of you for your time on this.
